src/extensions/idle.rs

```
@@ -108,39 +111,84 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
+    /// timeout duration set by [`Handle::timeout`]. If you do not want this behaviour, call
+    /// this function and the connection will simply IDLE until `wait_while` returns or
+    /// the timeout expires.
+    pub fn keepalive(&mut self, keepalive: bool) -> &mut Self {
+        self.keepalive = keepalive;
+        self
+    }
+
+    /// Block until the given callback returns `false`, or until a response
+    /// arrives that is not explicitly handled by [`UnsolicitedResponse`].
+    pub fn wait_while<F>(&mut self, mut callback: F) -> Result<WaitOutcome>
     where
         F: FnMut(UnsolicitedResponse) -> bool,
     {
         let mut v = Vec::new();
         let result = loop {
-            match self.session.readline(&mut v) {
+            match {
+                // The server MAY consider a client inactive if it has an IDLE command
+                // running, and if such a server has an inactivity timeout it MAY log
+                // the client off implicitly at the end of its timeout period.  Because
+                // of that, clients using IDLE are advised to terminate the IDLE and
+                // re-issue it at least every 29 minutes to avoid being logged off.
+                // This still allows a client to receive immediate mailbox updates even
+                // though it need only "poll" at half hour intervals.
+                let time_since_idle = Instant::now() - self.last_idle.map(Ok).unwrap_or_else(|| self.init())?;
+                if time_since_idle >= self.timeout {
+                    Err(Error::Io(io::ErrorKind::TimedOut.into()))
+                } else {
+                    self.session
+                        .stream
+                        .get_mut()
+                        .set_read_timeout(Some(self.timeout - time_since_idle))
+                        // set_read_timeout() can fail if the argument is Some(0), which can never
+                        // be the case here.
+                        .unwrap();
```

```suggestion
                        .expect("cannot be Some(0) since time is monotonically increasing");
```

---

Will it panic here on the first loop iteration when the `timeout` is set to `Duration::ZERO`? 
If so, would it be better to return an error instead?

---

An error is returned in that case
https://github.com/jonhoo/rust-imap/pull/303/files#diff-a11b127eccbbebaad79d0cfb2418302d979952456e70664a86161fc63b622de4R174

---

It seems that [`.unwrap_or_else(|| self.init().map(|()| Duration::ZERO))`](https://github.com/jonhoo/rust-imap/pull/303/files#diff-a11b127eccbbebaad79d0cfb2418302d979952456e70664a86161fc63b622de4R180) covers that `Err(Error::Io(io::ErrorKind::TimedOut.into()))` and leads to [`.set_read_timeout(Some(self.timeout - Duration::ZERO))`](https://github.com/jonhoo/rust-imap/pull/303/files#diff-a11b127eccbbebaad79d0cfb2418302d979952456e70664a86161fc63b622de4R186)

---

Right, on the first iteration `last_idle` is `None` so the check is skipped. I'm moving it to fix this issue.